index/scorch/segment/zap/merge.go

```diff
@@ -361,14 +349,27 @@ func persistMergedRest(segments []*SegmentBase, dropsIn []*roaring.Bitmap,
 
 		// update the field doc values
 		fdvEncoder := newChunkedContentCoder(uint64(chunkFactor), newSegDocCount-1, w, true)
-		for docNum, docTerms := range docTermMap {
-			if len(docTerms) > 0 {
-				err = fdvEncoder.Add(uint64(docNum), docTerms)
+
+		for segmentI, segment := range segmentsInFocus {
+			fieldIDPlus1 := uint16(segment.fieldsMap[fieldName])
+			if dvIter, exists := segment.fieldDvReaders[fieldIDPlus1-1]; exists &&
+				dvIter != nil {
+				err = dvIter.iterateAllDocValues(segment, func(docNum uint64, terms []byte) error {
+					if newDocNums[segmentI][docNum] == docDropped {
+						return nil
+					}
+					err := fdvEncoder.Add(newDocNums[segmentI][docNum], terms)
+					if err != nil {
+						return err
+					}
+					return nil
+				})
 				if err != nil {
 					return nil, 0, err
 				}
 			}
 		}
+
 		err = fdvEncoder.Close()
```

Comment:
one thought perhaps to address the issue of zap.Segment.VisitableDocValueFields()...

is perhaps if the fdvEncoder is empty, then don't do an fdvEncoder.Close() and don't do a final fdvEncoder.Write()?

instead, an empty fdvEncoder should have a fieldDvLocsStart/fieldDvLocsEnd value that tells a future reader that there's no info here.

of note, there might a be a subtle difference between "there's no info here" versus fieldDocValues are disabled...  not sure.  The subtlety might tell the reader in the future whether to do the cached-doc-values in-memory uninversion maneuver or not.